### ISSUE TITLE
sendmsg.c does not compile on Solaris

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5728]
|<img alt="toddmowen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| toddmowen reported|
|-|-|
|Trac ID|trac#5728|
|Type|defect|
|Created|2012-06-28 03:58:01Z|
|Branch|https://github.com/twisted/twisted/tree/solaris-build-5728|

We are trying to upgrade from version 11 to 12, on a Solaris 5.10 box with Python 2.7.

The Twisted-12.0.0 release builds without error. However, the later Twisted-12.1.0 release produces the following error during `python setup.py build`:

```
twisted/python/sendmsg.c: In function `sendmsg_sendmsg':
twisted/python/sendmsg.c:198: error: structure has no member named `msg_control'
twisted/python/sendmsg.c:199: error: structure has no member named `msg_controllen'
twisted/python/sendmsg.c:201: error: structure has no member named `msg_flags'
twisted/python/sendmsg.c:243: warning: implicit declaration of function `CMSG_SPACE'
twisted/python/sendmsg.c:268: error: structure has no member named `msg_control'
twisted/python/sendmsg.c:269: error: structure has no member named `msg_control'
twisted/python/sendmsg.c:274: error: structure has no member named `msg_control'
twisted/python/sendmsg.c:276: error: structure has no member named `msg_controllen'
twisted/python/sendmsg.c:285: warning: implicit declaration of function `CMSG_FIRSTHDR'
twisted/python/sendmsg.c:285: warning: initialization makes pointer from integer without a cast
twisted/python/sendmsg.c:306: warning: implicit declaration of function `CMSG_LEN'
twisted/python/sendmsg.c:316: warning: implicit declaration of function `CMSG_DATA'
twisted/python/sendmsg.c:316: warning: assignment makes pointer from integer without a cast
twisted/python/sendmsg.c:322: warning: implicit declaration of function `CMSG_NXTHDR'
twisted/python/sendmsg.c:322: warning: assignment makes pointer from integer without a cast
twisted/python/sendmsg.c:351: error: structure has no member named `msg_control'
twisted/python/sendmsg.c:352: error: structure has no member named `msg_control'
twisted/python/sendmsg.c:353: error: structure has no member named `msg_control'
twisted/python/sendmsg.c: In function `sendmsg_recvmsg':
twisted/python/sendmsg.c:414: error: structure has no member named `msg_control'
twisted/python/sendmsg.c:416: error: structure has no member named `msg_controllen'
twisted/python/sendmsg.c:429: warning: assignment makes pointer from integer without a cast
twisted/python/sendmsg.c:432: warning: assignment makes pointer from integer without a cast
twisted/python/sendmsg.c:450: warning: cast to pointer from integer of different size
twisted/python/sendmsg.c:477: error: structure has no member named `msg_flags'
error: command 'gcc' failed with exit status 1
```

It appears that on this platform it is necessary to define the `_XOPEN_SOURCE` and `_XOPEN_SOURCE_EXTENDED` macros in order to gain access to the `msg_control`, `msg_controllen`, and `msg_flags` members that the errors mention (according to http://stackoverflow.com/questions/1034587).

Attachments:

* [my-twisted-patch.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bbe/bbedcbbf334eeac567067748020a08d0c60ca748/d5bf14bfe4ce8fc7b9256a069dbe396aeaef28e5.patch) (507 bytes) - added by nibalizer on 2012-07-28 21:53:00Z - This patch allows twitsted to build on solaris.
* [twisted-5728-fix-build-on-sunos.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bbe/bbedcbbf334eeac567067748020a08d0c60ca748/1906d076e168d9f794d06f1aff1a97da85458d7b.patch) (626 bytes) - added by xaka on 2013-03-29 23:01:50Z - 
* [twisted-5728-fix-build-on-sunos.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bbe/bbedcbbf334eeac567067748020a08d0c60ca748/a1fd1e1f32378de3da7da9bae4ab8062d4fd36b4.patch) (1016 bytes) - added by xaka on 2013-03-29 23:11:56Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5728 5728
type__defect defect
reporter__toddmowen toddmowen
priority__normal normal
milestone__ 
branch__branches_solaris_build_5728 branches/solaris-build-5728
branch_author__itamarst itamarst
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1340855881000000 1340855881000000
changetime__1385230969000000 1385230969000000
version__None None
owner__itamar itamar
cc__xaka cc__zooko@...
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

Just a note about defects vs regressions.  Twisted is "supported" on a number of platforms.  Only on those platforms do we consider new defects in old functionality (such as building) "regressions".  You can find the list of supported platforms on [our continuous integration system](http://buildbot.twistedmatrix.com/boxes-supported?branch=trunk&num_builds=10).  There are various Windows, Linux, and even BSD configurations.  However, no Solaris.  If you're interested in Twisted on Solaris, consider helping out with that effort.

**Comment by twisted-trac** [comment]
|<img alt="nibalizer's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|nibalizer commented|
|-|-|

This patch should allow for Solaris to build. I'm also trying to learn the development process so if there are procedures I'm not following please let me know.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst commented|
|-|-|

I emailed the person who recently setup a Solaris build machine, but which is currently offline (http://buildbot.twistedmatrix.com/builders/solaris11-x86-py2.6). Hopefully he'll put it back up so we can test the patch.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst commented|
|-|-|

OK, the Solaris builder is up. It doesn't have a build error, but the sendmsg.so it builds does seem broken:

http://buildbot.twistedmatrix.com/builders/solaris11-x86-py2.6/builds/9/steps/select/logs/problems

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

Inferring from the slave name, we have a Solaris 11 builder?  This bug report seems to be about ''SunOS' 5.10 (as I understand it, there is no Solaris 5.10, but perhaps Solaris 10 is like SunOS 5.10?).  That seems like a significant version skew.  Though SunOS 5.10 was released 7 years ago, apparently support is planned for another 6 years by ... Oracle.  So perhaps someone could contribute a Solaris 10 slave to complement the Solaris 11 slave.

**Comment by twisted-trac** [comment]
|<img alt="nibalizer's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|nibalizer commented|
|-|-|

If buildbot can run as my unprivledged user at my university then I can set that up to run. I'll just nice all the processes and have to restart it when the system goes down for patching. 

SunOS 5.10 is Solaris 10.  Somewhere around SunOS 4 Sun renamed it Solairs. Uname continues to be SunOS. 

This is the full uname:

```
uname -a
SunOS chandra.cs.pdx.edu 5.10 Generic_147440-19 sun4v sparc SUNW,SPARC-Enterprise-T5220
```

**Comment by twisted-trac** [comment]
|<img alt="nibalizer's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|nibalizer commented|
|-|-|

While trying to easy_install buildbot I was unable to build twisted. Pasting here. Since buildbot depends on twisted... should I try with Twisted 12.0.0 ? Should I apply my patch?


```
Best match: Twisted 12.1.0
Downloading http://pypi.python.org/packages/source/T/Twisted/Twisted-12.1.0.tar.bz2#md5=f396f1d6f5321e869c2f89b2196a9eb5
Processing Twisted-12.1.0.tar.bz2
Running Twisted-12.1.0/setup.py -q bdist_egg --dist-dir /tmp/easy_install-hTDI_i/Twisted-12.1.0/egg-dist-tmp-Z2xXLG
conftest.c:1:23: sys/epoll.h: No such file or directory
twisted/python/sendmsg.c: In function `sendmsg_sendmsg':
twisted/python/sendmsg.c:198: error: structure has no member named `msg_control'
twisted/python/sendmsg.c:199: error: structure has no member named `msg_controllen'
twisted/python/sendmsg.c:201: error: structure has no member named `msg_flags'
twisted/python/sendmsg.c:243: warning: implicit declaration of function `CMSG_SPACE'
twisted/python/sendmsg.c:268: error: structure has no member named `msg_control'
twisted/python/sendmsg.c:269: error: structure has no member named `msg_control'
twisted/python/sendmsg.c:274: error: structure has no member named `msg_control'
twisted/python/sendmsg.c:276: error: structure has no member named `msg_controllen'
twisted/python/sendmsg.c:285: warning: implicit declaration of function `CMSG_FIRSTHDR'
twisted/python/sendmsg.c:285: warning: initialization makes pointer from integer without a cast
twisted/python/sendmsg.c:306: warning: implicit declaration of function `CMSG_LEN'
twisted/python/sendmsg.c:316: warning: implicit declaration of function `CMSG_DATA'
twisted/python/sendmsg.c:316: warning: assignment makes pointer from integer without a cast
twisted/python/sendmsg.c:322: warning: implicit declaration of function `CMSG_NXTHDR'
twisted/python/sendmsg.c:322: warning: assignment makes pointer from integer without a cast
twisted/python/sendmsg.c:351: error: structure has no member named `msg_control'
twisted/python/sendmsg.c:352: error: structure has no member named `msg_control'
twisted/python/sendmsg.c:353: error: structure has no member named `msg_control'
twisted/python/sendmsg.c: In function `sendmsg_recvmsg':
twisted/python/sendmsg.c:414: error: structure has no member named `msg_control'
twisted/python/sendmsg.c:416: error: structure has no member named `msg_controllen'
twisted/python/sendmsg.c:429: warning: assignment makes pointer from integer without a cast
twisted/python/sendmsg.c:432: warning: assignment makes pointer from integer without a cast
twisted/python/sendmsg.c:477: error: structure has no member named `msg_flags'
error: Setup script exited with error: command 'gcc' failed with exit status 1
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_9"></a>|@itamarst commented|
|-|-|

Any of those. Or just comment out the line in setup.py about sendmsg.c, it's not at all necessary.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

Thanks for the offer, but this doesn't sound like the kind of environment that is suited to be used as a build slave.  Slaves need to be reliably online 24/7.  Do you expect this user account on shared university resources will meet that requirement?

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_11"></a>|@itamarst commented|
|-|-|

Can anyone (besides original author of patch) verify that the patch fixes the build?

**Comment by twisted-trac** [comment]
|<img alt="dannygoulder's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|dannygoulder commented|
|-|-|

Confirmed this does fix the build:

-bash-3.2$ patch < ~goulderd/build/sendmsg.c.patch
  Looks like a unified context diff.
done
-bash-3.2$ bash ../mysetuppy.sh .
Processing .
Running setup.py -q bdist_egg --dist-dir /var/tmp/graphite/build/Twisted-12.2.0/egg-dist-tmp-N0owxZ
gcc: unrecognized option `-KPIC'
gcc: unrecognized option `-KPIC'
conftest.c:1:23: sys/epoll.h: No such file or directory
gcc: unrecognized option `-KPIC'
gcc: unrecognized option `-KPIC'
In file included from /usr/include/python2.6/Python.h:8,
                 from twisted/internet/_sigchld.c:9:
/usr/include/python2.6/pyconfig.h:978:1: warning: "_FILE_OFFSET_BITS" redefined
In file included from /usr/include/signal.h:18,
                 from twisted/internet/_sigchld.c:6:
/usr/include/sys/feature_tests.h:187:1: warning: this is the location of the previous definition
Adding Twisted 12.2.0 to easy-install.pth file
Installing pyhtmlizer script to /prod/systems/graphite/bin
Installing twistd script to /prod/systems/graphite/bin
Installing tap2deb script to /prod/systems/graphite/bin
Installing ckeygen script to /prod/systems/graphite/bin
Installing tap2rpm script to /prod/systems/graphite/bin
Installing conch script to /prod/systems/graphite/bin
Installing tapconvert script to /prod/systems/graphite/bin
Installing trial script to /prod/systems/graphite/bin
Installing tkconch script to /prod/systems/graphite/bin
Installing cftp script to /prod/systems/graphite/bin
Installing mailmail script to /prod/systems/graphite/bin
Installing manhole script to /prod/systems/graphite/bin
Installing lore script to /prod/systems/graphite/bin

Installed /prod/systems/graphite/lib/python2.6/site-packages/Twisted-12.2.0-py2.6-solaris-2.10-i86pc.egg
Processing dependencies for Twisted==12.2.0
Finished processing dependencies for Twisted==12.2.0
-bash-3.2$ uname -a
SunOS sysappsx1.london.daiwa.global 5.10 Generic_147441-24 i86pc i386 i86pc

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_13"></a>|@itamarst commented|
|-|-|

Sigh. The linked info suggests we want to set _XOPEN_SOURCE and _XOPEN_SOURCE_EXTENDED, but the patch sets something else, so we need a new patch setting those and then a retest.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

Also, still need a buildslave (otherwise effectively no one can work on this or test it).

**Comment by twisted-trac** [comment]
|<img alt="xaka's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_15"></a>|xaka commented|
|-|-|

Sorry for messing up with patches. The right one is twisted-5728-fix-build-on-sunos.2.patch.

**Comment by twisted-trac** [comment]
|<img alt="enmand's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_16"></a>|enmand commented|
|-|-|

If the only hold-up on this bug is that it needs a buildbot to test on it, I can offer a SmartOS buildbot for Python 2.6, 2.7, 3.2 and 3.3. Email me off-topic and we can organize it.

Thanks,
Dan

P.S. the patched worked for me.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_17"></a>|@itamarst commented|
|-|-|

enmand, could you email tom.prince@... about setting up a buildbot? 

Personally I'm fine accepting a patch given independent confirmation of compilation, but a buildbot would be much better if we want things to continue working.

Adding review keyword.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_18"></a>|@zooko set owner to xaka|
|-|-|

I reviewed [twisted-5728-fix-build-on-sunos.2.patch](None/twisted-5728-fix-build-on-sunos.2.patch) . Thank you, xaka, for contributing the patch!

Here are my comments:

 * The patch doesn't come with a NEWS entry. Please add a NEWS entry as described here: [ReviewProcess#Newsfiles](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles) so that Twisted users will be informed about the fix.
 * The comment seems to be some text from a man page, and it is describing what the X/Open CAE is in general instead of describing why Twisted needs to use X/Open CAE. Please replace that text with an explanation of why Twisted needs this, plus a reference to the X/Open CAE doc, for example like this:
 ```
# On Solaris 10, we need to define the _XOPEN_SOURCE and
# _XOPEN_SOURCE_EXTENDED macros to build in order to gain access to
# the msg_control, msg_controllen, and msg_flags members in
# sendmsg.c. (according to
# http://stackoverflow.com/questions/1034587).  See the documentation
# of X/Open CAE in the standards(5) man page of Solaris.
```

Otherwise the patch looks good to me. Please attach a new patch with these changes.

By the way, I did not apply the patch, build it (I don't have Solaris), run unit tests or pyflakes, use buildbot, or do any other steps besides read the patch and write the review comments above.

**Comment by twisted-trac** [comment]
|<img alt="s0undt3ch's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_19"></a>|s0undt3ch commented|
|-|-|

I confirm that [twisted-5728-fix-build-on-sunos.2.patch](None/twisted-5728-fix-build-on-sunos.2.patch) allowed me to compile 0.13.0 under SmartOS.

**Comment by twisted-trac** [comment]
|<img alt="s0undt3ch's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_20"></a>|s0undt3ch commented|
|-|-|

For those with an urgency in this fix, while it's not included into twisted:

```
pip install git+git://github.com/s0undt3ch/twisted.git#egg=Twisted
```

I just created this patched twisted clone because I need it for an automated install procedure.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun commented|
|-|-|

[#6787](https://github.com/twisted/twisted/issues/6787) was a duplicate of this.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_22"></a>|@itamarst commented|
|-|-|

(In [40559]) Branching to 'solaris-build-5728'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_23"></a>|@itamarst removed owner|
|-|-|

I've added news file, and improved setup.py comment. Started buildbot build.

My thought is that given passing tests on supported platforms, and people reporting this fixes the problem, I'm fine merging this. The news file explicitly mentions that Solaris is not a supported platform until we have a buildslave.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_24"></a>|@thijstriemstra set owner to @itamarst|
|-|-|

Thanks, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_25"></a>|@itamarst set status to `closed`|
|-|-|

(In [40859]) Merge solaris-build-5728.

Author: xaka
Reviewers: zooko, itamar, thijs
Fixes: [#5728](https://github.com/twisted/twisted/issues/5728)


sendmsg.c now builds on Solaris.
